{"nbformat":4,"nbformat_minor":0,"metadata":{"kernelspec":{"display_name":"R","language":"R","name":"ir"},"language_info":{"codemirror_mode":"r","file_extension":".r","mimetype":"text/x-r-source","name":"R","pygments_lexer":"r","version":"3.6.1"},"colab":{"name":"Data Analysis - R Notebook .ipynb","provenance":[],"collapsed_sections":[]}},"cells":[{"cell_type":"markdown","metadata":{"id":"mHoWOGkx5qcj"},"source":["# Data Analysis Notebook for Neuronal Differentation 2021\n","\n","Authored by Abhranil Maiti (Riju) - MSc Bioinformatics and Theoretical Systems Biology, Imperial College London\n","\n","After the gene counts have been created from the raw reads (see Bash notebook for Neuronal Differentation 2021 for more information), the next step is the data analysis and interpretation! \n","\n","For this notebook, you will need the following dependencies:\n","- R-Studio/R/R-kernel (if using this Jupyter Notebook)\n","- DESeq2\n","- ggplot2\n","\n","The files required for this notebook are:\n","- Read Count files (**NOT** TPMs)\n","- Metadata.txt file\n","\n","More information is found in Part 2\n","\n","This section will be split into four parts:\n","\n","- Part 1: Package installation (optional)\n","- Part 2: Data Preparation\n","- Part 3: DESeq2 Processing\n","- Part 4: Data Analysis\n","\n","If any part of the Jupyter Notebook appears to be out of date or is incorrect, or if there are any questions, please contact the author at: abhranil.maiti20@imperial.ac.uk, or on GitHub (riju.maiti)."]},{"cell_type":"markdown","metadata":{"id":"5oH-CZLy5qcx"},"source":["# Part 1: Package installation (optional)\n","\n","This step is optional if you do not have the required packages specified above. They can be installed by running the code-block/copying the code below.\n","\n","DESeq2 is part of the Bioconductor suite of programs. Thus, installation is quite straightforward as shown below:"]},{"cell_type":"code","metadata":{"id":"QpIooSe05qcz"},"source":["#checks if R/R-Studio has BiocManager already installed. If not, it will install BiocManager. \n","#Biocmanager is used for many other bioinformatics packages so it is useful to install\n","\n","if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n","    install.packages(\"BiocManager\")\n","BiocManager::install(version = \"3.12\")\n","\n","#note: version 3.12 is the latest as of 23/04/2021\n","\n","#using BiocManager, installs DeSeq2 on the machine. \n","\n","BiocManager::install(\"DESeq2\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"sVdhcLcp5qc0"},"source":["ggplot2 and ggpubr are also quite useful for plot visualisation, so it is also recommended to install these (will be used)"]},{"cell_type":"code","metadata":{"id":"fo33oWPD5qc1"},"source":["install.packages(\"ggplot2\")\n","install.packages(\"ggpubr\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"tqSgmYmyTgUW"},"source":["Don't forget to also load the libraries once installed!"]},{"cell_type":"code","metadata":{"id":"ZHnR-l6-Tlfw"},"source":["library(\"DESeq2\")\n","library(\"ggplot2\")\n","library(\"ggpubr\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"aYCe0A5q5qc2"},"source":["# Part 2: Data Preparation\n","Like with all data analysis software, there is an element of data preparation before it runs wild and generates some inferences. Here are the first few steps:"]},{"cell_type":"markdown","metadata":{"id":"lOZCT-At5qc3"},"source":["### **2.1** Preparing the raw counts file\n","\n","Lets first load the raw gene counts file, as shown here:"]},{"cell_type":"code","metadata":{"id":"k3IVBKOg5qc4"},"source":["#list your directory where your raw gene count files are located in the square brackets\n","#remember to remove the square brackets after!\n","loc_df <- \"[your-directory-here]\"\n","\n","#reads in the csv file into the dataframe\n","#if using a Tab Delimited File, use \"\\t\" instead of \",\" in the \"sep\" field\n","df <- read.table(loc_df, row.names=,1, sep=\",\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Q9Y9vyLw5qc6"},"source":["The Raw Gene Counts files should be obtained from the \"Bash notebook for Neuronal Differentation 2021\" notebook\n","\n","If we look inside this file however, we can see a problem:"]},{"cell_type":"code","metadata":{"id":"dAfYrufm5qdi"},"source":["head(df)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"rW1zhiyw5qdv"},"source":["The gene count files that have been created will contain the Ensembl IDs as well as the raw gene counts. Unfortunately, the Ensembl IDs will have their version numbers, which may cause confusion to the DESeq program downstream! To remedy this, we can use a simple R-script line as shown below:"]},{"cell_type":"code","metadata":{"id":"FoqSkfCQ5qdw"},"source":["#the term after the $ refers to the column that your ensembl_IDs are located in\n","#by default this should be titled \"gene_id\"\n","df$gene_id <- gsub(\"\\\\..*\",\"\",df$gene_id)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"EHUdQuglYMdT"},"source":["We can then save the files using the following command below. This was done for both mouse and human samples."]},{"cell_type":"code","metadata":{"id":"jysrlP6LYZfV"},"source":["write.csv(merged, \"[your_counts_file].csv\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"plT4Zrue5qdx"},"source":["### **2.2** Preparing the metadata.txt file\n","\n","With every experiment, there are usually several replicates. However, we need to specify to DESeq which samples are biological replicates, and which samples are replicates for different conditions! \n","\n","The format for the metadata.txt file is quite easy. Using our data for instance, they were laid out as follows (mouse):\n","\n","Day 0: m_0_1, m_0_2, m_0_3\n","\n","Day 0.5: m_0.5_1, m_0.5_2, m_0.5_3\n","\n","Day 1: m_1_1, m_1_2, m_1_3\n","etc. \n","\n","Therefore, the metadata.txt file would look as follows:\n","\n","<br>sample &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; group\n","<br>m_0_1  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  day_0\n","<br>m_0_2  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  day_0\n","<br>m_0_3  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  day_0\n","<br>m_0.5_1 &nbsp; &nbsp; &nbsp; day_0.5\n","<br>m_0.5_2 &nbsp; &nbsp; &nbsp; day_0.5\n","<br>m_0.5_3 &nbsp; &nbsp; &nbsp; day_0.5\n","<br>m_1_1 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;   day_1\n","<br>m_1_2 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;   day_1\n","<br>m_1_3 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;   day_1\n","<br>etc. \n","\n","The metadata.txt files can be created using Microsoft Excel, and saved as a \"Text (Tab Delimited) (.txt)\". This option is proposed in Microsoft Excel\n","\n","**IMPORTANT** You cannot have multiple spaces between words when creating your groups/samples. In the example above, \"Day 0\" would need to be converted to \"Day_0\".\n","\n","**IMPORTANT** Your metadata.txt file must be ordered in the same way that your headings are saved as. For instance, an excerpt of our human data headings were as follows:\n","\n","... h_11_2   h_11_3   h_1_1   h_12_1  h_12_2  h_12_3  h_1_2 ...\n","\n","Therefore, the metadata needed to be structured as follows:\n","\n","<br>sample &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; group\n","<br>h_11_2 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; day 11\n","<br>h_11_3 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; day 11\n","<br>h_1_1 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; day 1\n","<br>h_12_1 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; day 12\n","<br>h_12_2 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; day 12\n","<br>h_12_3 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; day 12\n","<br>h_1_2 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; day 1\n","\n","After doing that, you can simply load the metadata.txt file into R"]},{"cell_type":"code","metadata":{"id":"T2cDra3D5qd1"},"source":["#list your directory where your metadata.txt file is located in the square brackets\n","#remember to remove the square brackets after!\n","loc_metadata <- \"[your-directory-here]\"\n","\n","#reads in the metadata file into the dataframe\n","metadata <- read.table(loc_metadata, row.names=,1) "],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"sU_0olBYkS1e"},"source":["For our DESeq2 analysis, we combined the mouse and human datasets together. Therefore, in the metadata, you need another column titled \"species\" so that the program can distinguish the samples that came from mouse, and the samples that came from human. See below for example:\n","\n","<br>sample &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; group &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; species\n","<br>m_0_1  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  day_0 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; mouse\n","<br>m_0_2  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  day_0 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; mouse\n","<br>m_0_3  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  day_0 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; mouse\n","<br>h_12_1 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; day 12 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; human\n","<br>h_12_2 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; day 12 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; human\n","<br>h_12_3 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; day 12 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; human\n","\n","Don't forget, the metadata must be ordered in the same way the headings that are found in the file!"]},{"cell_type":"markdown","metadata":{"id":"ySrl0wSpW58d"},"source":["### **2.3** Orthologous Gene Extraction\n","\n","Next, we extracted orthologous genes so that sample comparisons between mouse and humans. This is optional for other analysis, but important for species comparisons. The code below was used. \n","\n","NOTE: List of Orthologs was obtained from BioMart (https://www.ensembl.org/biomart/martview/) and the following fields selected: \"Gene Stable ID\", \"Gene Name\""]},{"cell_type":"code","metadata":{"id":"aigsIIrXXNM_"},"source":["#inputting the files saved earlier for both mouse and human counts\n","df_human <- read.csv(\"[file_containing_human_counts.tsv\", sep=\"\\t\", header=TRUE)\n","df_mouse <- read.csv(\"[filee_containing_mouse_counts.tsv\", sep = \"\\t\", header=TRUE)\n","\n","#inputting the files detailing the mouse orthologs to human genes, and vice versa\n","#two DISTINCT files are needed as one file will have human ENSEMBL gene IDs\n","#the other file will have mouse ENSEMBL IDs\n","df_orthologs_mouse <- read.csv(\"human_mouse_orthologs.txt\", sep=\"\\t\", header=TRUE)\n","df_orthologs_human <- read.csv(\"mouse_human_orthologs.txt\", sep=\"\\t\", header=TRUE)\n","\n","#the ortholog gene list is merged ONLY if there are matching gene IDs in both files\n","#gene IDs that are not found in either list are omitted.\n","#if the final dataframe output has nothing in it, check that the orthologs list you are merging\n","#hasn't been switched\n","\n","merged_human <- merge(df_human, data_orthologs_human, by.x=\"Gene.stable.ID\", by.y=\"Gene\", all=FALSE)\n","merged_mouse <- merge(df_mouse, data_orthologs_mouse, by.x=\"Gene.stable.ID\", by.y=\"Gene\", all=FALSE)\n","\n","#next, both mouse and human files were to make one big file containing both samples:\n","#for the merge to be successful, the mouse gene names need to be capitalised\n","merged_master <- merge(df_human, data_mouse, by.x=\"Gene.Name\", by.y=\"Gene.Name\", all=FALSE)\n","\n","#files can then be written to a csv file after\n","write.csv(merged_master, \"DESEq2_ready_file.csv\")\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"JNnp47qj5qd2"},"source":["# Part 3: DESeq2 processing\n","\n","Once the files have been setup and loaded, DESeq2 can be run.\n","\n","Note: Prior to this step, the ENSEMBL ID columns were deleted. The remaining columns should just be one column with gene names, and the samples containing raw gene counts.\n","\n","Before doing so, we can perform an (optional) step of excluding genes whos counts are less than 10 overall, or genes that have no recorded expression. This can be done by executing the code below:"]},{"cell_type":"code","metadata":{"id":"sHtOK-135qd3"},"source":["#removes genes who do not have any recorded expression at all.\n","#you can change 0 to be anything i.e. 10 to only include genes that have a summed count of 10 or more for a particular gene.\n","merged_master <- merged_master[rowSums(merged_master[, -1])>0, ]"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"601Jl9W75qd7"},"source":["This step isn't always necessary, but it does speed up the processing time for DESeq2 in case you do have a large number of samples."]},{"cell_type":"markdown","metadata":{"id":"fyyL2a3LVq8r"},"source":["Next, a DESeq2 matrix is created. This is done as follows:"]},{"cell_type":"code","metadata":{"id":"44Ep5ICK5qd9"},"source":["#creates a summarised DESeq2 matrix\n","dds <- DESeqDataSetFromMatrix(countData=merged_master, colData=metadata, design=~group+species)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"vMwHlMY-5yjl"},"source":["We then need to run DESeq2:"]},{"cell_type":"code","metadata":{"id":"9B8OJm9553KR","executionInfo":{"status":"aborted","timestamp":1619208058232,"user_tz":-60,"elapsed":1403,"user":{"displayName":"Riju Maiti","photoUrl":"","userId":"16640262794535149576"}}},"source":["#command to run the differential gene expression analysis\n","dds <- DESeq(dds)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"gNDGzbF45qeA"},"source":["As we are concerned with doing network analysis, we will only be using DESeq2 to generate homoskedastic data i.e. variance does not depend on the mean. We need to do this as in RNASeq data, the mean grows with the variance. The code is as below:"]},{"cell_type":"code","metadata":{"id":"8lbOVWSt5qeA","executionInfo":{"status":"aborted","timestamp":1619208058233,"user_tz":-60,"elapsed":1400,"user":{"displayName":"Riju Maiti","photoUrl":"","userId":"16640262794535149576"}}},"source":["newdata <- getVarianceStabilizedData(dds) "],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"duQvH5ZH5qeF"},"source":["We can then save our normalised data as shown here:"]},{"cell_type":"code","metadata":{"id":"e-AQ_byJ5qeG","executionInfo":{"status":"aborted","timestamp":1619208058235,"user_tz":-60,"elapsed":1400,"user":{"displayName":"Riju Maiti","photoUrl":"","userId":"16640262794535149576"}}},"source":["#writes a csv file of your variance stabilized data. Change the file name as is required. \n","write.csv(newdata, \"[your file name here]\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"Dm3agcTzUf5J"},"source":["# Part 4: Data Analysis\n","\n","Now for the fun part; data analysis! This is where you can finally view just how pretty the data is.\n","\n","This section is split into three parts (based on the project):\n","\n","*   Generating Plots\n","*   Principal Component Analysis\n","*   Time-Factor Estimations\n","\n","\n","\n"]},{"cell_type":"markdown","metadata":{"id":"anBDK7IXcajk"},"source":["### **3.1** Generating Plots\n","\n","For this section, we need the following packages: ggplot2, and ggpubr.\n","\n","ggplot2 does the heavy lifting in terms of plot generation. \n","ggpubr arranges the plots nicely for publications.\n","\n","Prior to running the code, Pax6, Irx3, Nkx2.2, and Olig2.2 were extracted using Excel, data feature-scaled (normalised between 0 and 1) and then values were averaged. The file format was as follows:\n","\n","<br>day &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Pax6 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Olig2 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Nkx2.2 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; Irx3\n","<br>0.5  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  0.00286 &nbsp; &nbsp; &nbsp;\n","0.00024 &nbsp; &nbsp;\n","0.00010 &nbsp; &nbsp; &nbsp; &nbsp;\n","0.0723\n","<br>1  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;  0.84609  &nbsp; &nbsp; &nbsp;\n","0.22127 &nbsp; &nbsp;\n","0.01384 &nbsp; &nbsp; &nbsp; &nbsp;\n","0.89926\n","<br>2  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; 0.49591 &nbsp; &nbsp; &nbsp;\n","0.54934 &nbsp; &nbsp; \n","0.04497 &nbsp; &nbsp; &nbsp; &nbsp;\n","0.82328\n","<br> etc.\n","\n","\n","For TPM vs VST data, the following code was used:"]},{"cell_type":"code","metadata":{"id":"gXJtJtEPfxrQ"},"source":["#example code for reading in a file containing processed data as described above.\n","df_mouse_TPM <- read.csv(\"[data_for_mouse_in_TPM.csv]\", sep=\",\", header=TRUE)\n","df_mouse_VST <- read.csv(\"[data_for_mouse_in_VST.csv]\", sep=\",\", header=TRUE)\n","\n","df_human_TPM <- read.csv(\"[data_for_human_in_TPM.csv]\", sep=\",\", header=TRUE)\n","df_human_VST <- read.csv(\"[data_for_human_in_VST.csv]\", sep=\",\", header=TRUE)\n","\n","#breaking down an example\n","TPM_mouse <- ggplot(df_mouse_TPM, aes(x = Day)) + #defines a new ggplot using mouse_TPMs\n","  geom_line(aes(y = Pax6, color = \"Pax6\"), size = 1.2) + #defines the Pax6 line using Pax6 data from the dataframe\n","  geom_line(aes(y = Nkx2.2, color = \"Nkx2.2\"), size = 1.2) + #defines the Nkx2.2 line using Pax6 data from the dataframe\n","  geom_line(aes(y = Olig2 color = \"Olig2\"), size = 1.2) + #defines the Olig2 line using Pax6 data from the dataframe\n","  geom_line(aes(y = Irx3, color = \"Irx3\"), size = 1.2) + #defines the Irx3 line using Pax6 data from the dataframe\n","  ggtitle(\"Mouse GRN (TPM)\") + #creates a title for the plot\n","  labs(x = \"Days in RA + SAG\", y = \"Normalised Expression\", size = 15) + #labels the axes\n","  xlim(0, 7) + #sets the limit of the x-axis\n","  \n","  theme(axis.line = element_line(\"black\"), panel.background = element_rect(fill = 'white'),  #changes the background to be white and the axis lines grey\n","        legend.title=element_blank(), legend.text = element_text(size = 15), #removes the legend title, and makes the legend text size 15\n","        plot.title = element_text(size=15, face=\"bold.italic\", hjust = 0.5), #makes the plot title 15, bold.italic, and centres the title\n","        text = element_text(size = 15)) + #makes the axis text size 15\n","  \n","  scale_colour_manual(values=c(\"#2ca02c\", \"#1f77b4\", \"#d62728\", \"#ff7f0e\")) #sets the colour of the Pax6, Irx3, Olig2, and Nk2.2 lines\n","\n","TPM_human <- ggplot(df_human_TPM, aes(x = Day)) +\n","  geom_line(aes(y = Pax6, color = \"Pax6\"), size = 1.2) +\n","  geom_line(aes(y = Nkx2.2, color = \"Nkx2.2\"), size = 1.2) +\n","  geom_line(aes(y = Olig2, color = \"Olig2\"), size = 1.2) +\n","  geom_line(aes(y = Irx3, color = \"Irx3\"), size = 1.2) +\n","  ggtitle(\"Human GRN (TPM)\") +\n","  labs(x = \"Days in RA + SAG\", y = \"Normalised Expression\", size = 15) +\n","  xlim(0, 15) +\n","  \n","  theme(axis.line = element_line(\"black\"), panel.background = element_rect(fill = 'white'), \n","        legend.title=element_blank(), legend.text = element_text(size = 15),\n","        plot.title = element_text(size=15, face=\"bold.italic\", hjust = 0.5), \n","        text = element_text(size = 15)) +\n","  \n","  scale_colour_manual(values=c(\"#2ca02c\", \"#1f77b4\", \"#d62728\", \"#ff7f0e\"))\n","\n","VST_mouse <- ggplot(df_mouse_VST, aes(x = Day)) +\n","  geom_line(aes(y = Pax6, color = \"Pax6\"), size = 1.2) +\n","  geom_line(aes(y = Nkx2.2, color = \"Nkx2.2\"), size = 1.2) +\n","  geom_line(aes(y = Olig2, color = \"Olig2\"), size = 1.2) +\n","  geom_line(aes(y = Irx3, color = \"Irx3\"), size = 1.2) +\n","  ggtitle(\"Mouse GRN (VST)\") +\n","  labs(x = \"Days in RA + SAG\", y = \"Normalised Expression\", size = 15) +\n","  xlim(0, 7) +\n","  \n","  theme(axis.line = element_line(\"black\"), panel.background = element_rect(fill = 'white'), \n","        legend.title=element_blank(), legend.text = element_text(size = 15),\n","        plot.title = element_text(size=15, face=\"bold.italic\", hjust = 0.5), \n","        text = element_text(size = 15)) +\n","  \n","  scale_colour_manual(values=c(\"#2ca02c\", \"#1f77b4\", \"#d62728\", \"#ff7f0e\"))\n","\n","VST_human <- ggplot(df_human_VST, aes(x = Day)) +\n","  geom_line(aes(y = Pax6, color = \"Pax6\"), size = 1.2) +\n","  geom_line(aes(y = Nkx2.2, color = \"Nkx2.2\"), size = 1.2) +\n","  geom_line(aes(y = Olig2, color = \"Olig2\"), size = 1.2) +\n","  geom_line(aes(y = Irx3, color = \"Irx3\"), size = 1.2) +\n","  ggtitle(\"Human GRN (VST)\") +\n","  labs(x = \"Days in RA + SAG\", y = \"Normalised Expression\", size = 15) +\n","  xlim(0, 15) +\n","  \n","  theme(axis.line = element_line(\"black\"), panel.background = element_rect(fill = 'white'), \n","        legend.title=element_blank(), legend.text = element_text(size = 15),\n","        plot.title = element_text(size=15, face=\"bold.italic\", hjust = 0.5), \n","        text = element_text(size = 15)) +\n","  \n","  scale_colour_manual(values=c(\"#2ca02c\", \"#1f77b4\", \"#d62728\", \"#ff7f0e\"))\n","\n","#this function from ggpubr arranges the plots together, and combines the legend and displays it at the bottom of the plots\n","  ggarrange(TPM_human, VST_human, TPM_mouse, VST_mouse, labels = c(\"A\", \"B\", \"C\", \"D\"), common.legend = TRUE, legend = \"bottom\")\n"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"hCbcQBpHjlNc"},"source":["### **3.2** Principal Component Analysis\n","For this section, we need the following packages: ggplot2, DESeq2\n","\n","We can just use the Variance Stabilised Data that had been created previously, as well as the following code."]},{"cell_type":"code","metadata":{"id":"BWJa-tjWmArr"},"source":["pcaData <- plotPCA(newdata, intgroup=c(\"group\", \"species\"), returnData=TRUE) #group is specified depending on the headings in the metadata.txt\n","#this function calculates the variance of each gene in the dataset before doing the dimensionality reduction\n","\n","ggplot(pcaData, aes(PC1, PC2, color=group.1, shape=species)) + #creates the plot using PC1 and PC2 found in \"pcaData\"\n","  geom_point(size=3) + #sets the size  of the data points\n","  xlab(paste0(\"PC1: \",percentVar[1],\"% variance\")) + #sets the x-label\n","  ylab(paste0(\"PC2: \",percentVar[2],\"% variance\")) + #sets the y-label\n","  theme(axis.line = element_line(\"black\"), panel.background = element_rect(fill = 'white')) + #sets the background\n","  coord_fixed()\n","\n","  #additionally, the plots should automatically generate the legends, and represent the different species and time-points using different shapes/colours\n","  #this is to distinguish the samples from one another.\n","\n","#writes a csv file of PCA data for both mouse and human. Change the file name as is required. \n","write.csv(pcaData, \"[your file name here]\")"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"LQL7ahUio1RM"},"source":["### **3.3** Time Factor Estimations\n","For this section, we need the following packages: ggplot2, DESeq2\n","\n","Firstly, a file comparing PC2 and Days (of experiment) was prepared in Excel for mouse and human data. Format was below:\n","\n","<br>Day &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; PC2 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; species\n","<br>0 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; -40.20901 &nbsp; &nbsp;  human \n","<br>0 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; -40.54826 &nbsp; &nbsp;  human \n","<br>0 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; -39.96980 &nbsp; &nbsp;  human \n","<br>1 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; -32.17217 &nbsp; &nbsp;  human \n","<br>1 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; -32.31412 &nbsp; &nbsp;  human \n","<br>1 &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; -32.01572 &nbsp; &nbsp;  human \n","<br> etc.\n","\n","Then, the regression analysis was carried out as follows:"]},{"cell_type":"code","metadata":{"id":"WDDcTkmuqJ2O"},"source":["#reads in the dataframes for both mouse and human\n","df <- read.csv(\"[data_for_mouse_human_PC2.csv]\", sep=\",\", header=TRUE)\n","\n","df_mouse <- df[49:81, ] #extracting only mouse data\n","df_human <- df[4:45, ] #extracting only human data\n","\n","#IMPORTANT: 0 was excluded from regression analysis\n","\n","#assigns variables for regression analysis\n","x.mouse <- df_mouse$Day\n","y.mouse <- df_mouse$PC2.\n","\n","x.human <- df_human$Day\n","y.human <- df_human$PC2.\n","\n","#runs the regression analysis\n","model.mouse <- lm(y.mouse ~ log(x.mouse))\n","summary(model.mouse)\n","\n","model.human <- lm(y.human ~ log(x.human))\n","summary(model.human)"],"execution_count":null,"outputs":[]},{"cell_type":"markdown","metadata":{"id":"6s7MQYYqr7iD"},"source":["From the regression analysis, the data points could then be plotted, and the regression curves obtained from the regression analysis can also be overlayed."]},{"cell_type":"code","metadata":{"id":"y7LQn1aUqJu0"},"source":["fun.1 <- function(x) 28.198*log(x) - 20.246 #equation based off the regression analysis run for mouse data\n","fun.2 <- function(x) 19.8707*log(x) - 37.626 #equation based off the regression analysis run for human data\n","\n","#functions were then plotted\n","ggplot(df, aes(x = Day, y = PC2..Time., group = species, color = species)) + #creates initial plots\n","  geom_point(stat=\"identity\") + #adds each datapoint to the graph\n","  labs(x = \"Days in RA + SAG\", y = \"Gene Expression Profile (PC2)\") + #sets the labels of the axes\n","  theme(axis.line = element_line(\"black\"), panel.background = element_rect(fill = 'white'), #makes the background white and axes black\n","        legend.title=element_blank(), legend.text = element_text(size = 16), axis.title=element_text(size=16), #sets the size of the text\n","     axis.text=element_text(size=16)) +\n","\n","  geom_function(fun = fun.2, color=\"#F8766D\", size = 1.2) + #sets the colour of the mouse function\n","  geom_function(fun = fun.1, color=\"#00BFC4\", size = 1.2) + #sets the colour of the human function\n","  xlim(0.5, 15) #sets the limit of the graphs"],"execution_count":null,"outputs":[]}]}